components/VM_Arm/src/modules/virtio_con.c

```diff
@@ -22,61 +22,117 @@
 #include <platsupport/serial.h>
 #include <virtio/virtio_console.h>
 
-static virtio_con_t *virtio_con = NULL;
-extern void *serial_getchar_buf;
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
 
-//this matches the size of the buffer in serial server
-#define BUFSIZE 4088
+#define NUM_PORTS ARRAY_SIZE(serial_layout)
 
-char txbuf[BUFSIZE];
+static virtio_con_t *virtio_con = NULL;
 extern vmm_pci_space_t *pci;
 extern vmm_io_port_list_t *io_ports;
 
-struct {
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE 4088
+
+/* This struct occupies exactly 1 page and represents the data in the shmem region */
+typedef struct serial_shmem {
     uint32_t head;
     uint32_t tail;
     char buf[BUFSIZE];
-} extern volatile *batch_buf;
+} serial_shmem_t;
+compile_time_assert(serial_shmem_1k_size, sizeof(serial_shmem_t) == 0x1000);
 
-static int handle_serial_console(vm_t *vmm, void *cookie UNUSED)
-{
-    struct {
-        uint32_t head;
-        uint32_t tail;
-        char buf[BUFSIZE];
-    } volatile *buffer = serial_getchar_buf;
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct serial_conn {
+    void (*notify)(void);
+    volatile serial_shmem_t *recv_buf;
+    volatile serial_shmem_t *send_buf;
+} serial_conn_t;
+static serial_conn_t connections[NUM_PORTS];
+
+/* This is a temporary buffer used to stage data before completing the virtio RX */
+char temp_rx_buf[BUFSIZE];
 
+void virtio_console_putchar(int port, virtio_emul_t *con, char *buf, int len) WEAK;
+static int virtio_con_notify_recv(int port, virtqueue_device_t *queue)
+{
+    volatile serial_shmem_t *buffer = connections[port].recv_buf;
     int count = 0;
     while (buffer->head != buffer->tail) {
-        txbuf[count % BUFSIZE] = buffer->buf[buffer->head];
+        temp_rx_buf[count] = buffer->buf[buffer->head];
         buffer->head = (buffer->head + 1) % BUFSIZE;
-        count += 1 % BUFSIZE;
+        count++;
```

Comment:
This could overrun the buffer as txbuf[count] chases head.  It should also be possible to rework this to be multicore safe.
And again, do we absolutely need to do an extra copy here?

Reply:
temp_rx_buf is BUFSIZE bytes big. the shmem buffer is also BUFSIZE bytes big, so as long as both buffers are the same size, we can just check that the shmem head is not at the tail.

The extra copy can be avoided if the virtio emul layer knows that the shmem area is a circular buffer.

Reply:
Make it so ... make the virtIOemul layers understand the shape of the buffers it is dealing with.